contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -3910,3 +3912,494 @@ update_db_owner(const char *new_owner_name, const char *db_name)
+		values[LOGIN_EXT_DEFAULT_DATABASE_NAME] = CStringGetTextDatum(new_db_name);
+		replaces[LOGIN_EXT_DEFAULT_DATABASE_NAME] = true;
+
+		new_tuple = heap_modify_tuple(old_tuple,
+									bbf_authid_login_ext_dsc,
+									values,
+									nulls,
+									replaces);
+
+		CatalogTupleUpdate(bbf_authid_login_ext_rel, &old_tuple->t_self, new_tuple);
+		heap_freetuple(new_tuple);
+	}
+	systable_endscan(scan);
+
+	/* Close bbf_authid_login_ext, but keep lock till commit. */
+	table_close(bbf_authid_login_ext_rel, RowExclusiveLock);
+}
+
+static char
+*gen_rename_schema_or_role_cmds(char *old_name, char *new_name, bool is_schema)
+{
+	StringInfoData query;
+	initStringInfo(&query);
+	/*
+	 * We prepare the following query to Rename a schema or a ROLE.
+	 *
+	 * ALTER SCHEMA/ROLE <name>
+	 *
+	 */
+	appendStringInfo(&query, is_schema ? "ALTER SCHEMA " : "ALTER ROLE ");
+
+	appendStringInfo(&query, "%s RENAME TO %s", old_name, new_name);
+	return query.data;
+}
+
+/*
+ * We use processUtility to execute the renaming of schemas and roles
+ * instead of directly calling the individual APIs to rename them in
+ * order to adhere to the common high level code flow on which a lot of
+ * features depend like Beegees.
+ */
+static void
+exec_rename_db_util(char *old_db_name, char *new_db_name, bool is_schema)
+{
+	char *query_str = gen_rename_schema_or_role_cmds(old_db_name, new_db_name, is_schema);
+	List		*res;
+	Node	   	*res_stmt;
+	PlannedStmt *wrapper;
+
+	/*
+	 * The above query will be
+	 * executed using ProcessUtility()
+	 */
+	res = raw_parser(query_str, RAW_PARSE_DEFAULT);
+	res_stmt = ((RawStmt *) linitial(res))->stmt;
+
+	/* need to make a wrapper PlannedStmt */
+	wrapper = makeNode(PlannedStmt);
+	wrapper->commandType = CMD_UTILITY;
+	wrapper->canSetTag = false;
+	wrapper->utilityStmt = res_stmt;
+	wrapper->stmt_location = 0;
+	wrapper->stmt_len = 0;
+
+	ProcessUtility(wrapper,
+				is_schema ? "(RENAME SCHEMA )" : "(RENAME ROLE )",
+				false,
+				PROCESS_UTILITY_SUBCOMMAND,
+				NULL,
+				NULL,
+				None_Receiver,
+				NULL);
+
+	pfree(query_str);
+}
+
+void
+rename_tsql_db(char *old_db_name, char *new_db_name)
+{
+	int xactStarted = IsTransactionOrTransactionBlock();
+	Oid save_userid = InvalidOid;
+	int save_sec_context = 0;
+	int dbid = get_db_id(old_db_name);
+	int tries;
+	Oid     	prev_current_user = InvalidOid;
+
+	/* Alter database is not allowed inside a transaction. */
+	PreventInTransactionBlock(true, "ALTER DATABASE");
+
+	/*
+	 * Check that db_name is not "master", "tempdb", or "msdb",
+	 * IDs 1-4 are reserved for these native system databases.
+	 */
+	if (dbid == 0)
+		ereport(ERROR,
+			(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				errmsg("Database %s does not exist. Make sure that the name is entered correctly.", old_db_name)));
+	if (dbid && dbid <= 4)
+		ereport(ERROR,
+			(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				errmsg("Cannot change the name of the system database %s.", old_db_name)));
```

Comment:
Test cases for this condition are missing.

Reply:
Looks like the test cases are added but there was fault in the syntax, I will update it